### PR TITLE
acc-related fixes and cleanups

### PR DESCRIPTION
**Pull request by bartekgorny**
In some corner cases (e.g. when using pubsub with stream management on) it triggered errors on hook calls, because some handlers rely on server key being present (which is usually the case). It produced error log messages, made some metrics incorrect, and could potentially cause more trouble.

I think we should make those two values required in accumulator, as kind of "environment variables", whereby server is host in context of which we are operating, and user is the caller. Server-generated actions and admin's rest api calls might have user=undefined.

Also, this PR fixes acc processing in iq handling machinery, also cleaning it up so that functions have the same signatures and fixing function specs.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[4905.1](https://travis-ci.org/esl/MongooseIM/jobs/391712459) / Erlang 19.3 / internal_mnesia / 46dc9c1bbb6b9027475350dc0fd47c96b590b57a
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1925/4905/internal_mnesia.19.3) / [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4905/internal_mnesia.19.3/big/ct_run.test%40travis-job-79031576-d3a1-4500-aa93-460b12557e8b.2018-06-13_12.11.48/index.html) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4905/internal_mnesia.19.3/small/ct_run.mongooseim@localhost.2018-06-13_12.04.17/lib.mongooseim.logs/run.2018-06-13_12.04.18/suite.log.html)
**OK: 1069** / **Failed: 1** / **User-skipped: 41** / Auto-skipped: 0 
<details><summary>pep_SUITE:pep_tests:delayed_receive_with_sm</summary>


```erlang
{error,{{badmatch,1},
    [{pep_SUITE,'-delayed_receive_with_sm/1-fun-0-',3,
          [{file,"pep_SUITE.erl"},{line,197}]},
     {escalus_story,story,4,
            [{file,"/home/travis/build/esl/MongooseIM/big_tests/_build/default/lib/escalus/src/escalus_story.erl"},
             {line,72}]},
     {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1529}]},
     {test_server,run_test_case_eval1,6,
            [{file,"test_server.erl"},{line,1045}]},
     {test_server,run_test_case_eval,9,
            [{file,"test_server.erl"},{line,977}]}]}}
```

[Report log](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4905/internal_mnesia.19.3/big/ct_run.test%40travis-job-79031576-d3a1-4500-aa93-460b12557e8b.2018-06-13_12.11.48/big_tests.tests.pep_SUITE.logs/run.2018-06-13_12.23.45/pep_suite.delayed_receive_with_sm.1886.html)
</details>

**Comment by mongoose-im** [comment]
[4911.1](https://travis-ci.org/esl/MongooseIM/jobs/391815567) / Erlang 19.3 / internal_mnesia / c2dd64545c10730037568ea83ad7c5fa381ad33a
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1925/4911/internal_mnesia.19.3) / [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4911/internal_mnesia.19.3/big/ct_run.test%40travis-job-25d42030-35ea-42f8-b82b-bd2d10897c78.2018-06-13_16.23.02/index.html) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4911/internal_mnesia.19.3/small/ct_run.mongooseim@localhost.2018-06-13_16.16.13/lib.mongooseim.logs/run.2018-06-13_16.16.13/suite.log.html)
**OK: 1069** / **Failed: 1** / **User-skipped: 41** / Auto-skipped: 0 
<details><summary>pep_SUITE:pep_tests:delayed_receive_with_sm</summary>


```erlang
{error,{{badmatch,1},
    [{pep_SUITE,'-delayed_receive_with_sm/1-fun-0-',3,
          [{file,"pep_SUITE.erl"},{line,197}]},
     {escalus_story,story,4,
            [{file,"/home/travis/build/esl/MongooseIM/big_tests/_build/default/lib/escalus/src/escalus_story.erl"},
             {line,72}]},
     {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1529}]},
     {test_server,run_test_case_eval1,6,
            [{file,"test_server.erl"},{line,1045}]},
     {test_server,run_test_case_eval,9,
            [{file,"test_server.erl"},{line,977}]}]}}
```

[Report log](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4911/internal_mnesia.19.3/big/ct_run.test%40travis-job-25d42030-35ea-42f8-b82b-bd2d10897c78.2018-06-13_16.23.02/big_tests.tests.pep_SUITE.logs/run.2018-06-13_16.34.47/pep_suite.delayed_receive_with_sm.1467.html)
</details>

**Comment by mongoose-im** [comment]
[4914.1](https://travis-ci.org/esl/MongooseIM/jobs/391850324) / Erlang 19.3 / internal_mnesia / f83a0413ccf738eb22d9f5b304fd6e1f5affa5eb
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1925/4914/internal_mnesia.19.3) / [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4914/internal_mnesia.19.3/big/ct_run.test%40travis-job-2a97520b-2981-4cfe-b683-acf6e27dadc5.2018-06-13_17.34.05/index.html) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4914/internal_mnesia.19.3/small/ct_run.mongooseim@localhost.2018-06-13_17.27.13/lib.mongooseim.logs/run.2018-06-13_17.27.13/suite.log.html)
**OK: 1069** / **Failed: 1** / **User-skipped: 41** / Auto-skipped: 0 
<details><summary>pep_SUITE:pep_tests:delayed_receive_with_sm</summary>


```erlang
{error,{{badmatch,1},
    [{pep_SUITE,'-delayed_receive_with_sm/1-fun-0-',3,
          [{file,"pep_SUITE.erl"},{line,195}]},
     {escalus_story,story,4,
            [{file,"/home/travis/build/esl/MongooseIM/big_tests/_build/default/lib/escalus/src/escalus_story.erl"},
             {line,72}]},
     {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1529}]},
     {test_server,run_test_case_eval1,6,
            [{file,"test_server.erl"},{line,1045}]},
     {test_server,run_test_case_eval,9,
            [{file,"test_server.erl"},{line,977}]}]}}
```

[Report log](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4914/internal_mnesia.19.3/big/ct_run.test%40travis-job-2a97520b-2981-4cfe-b683-acf6e27dadc5.2018-06-13_17.34.05/big_tests.tests.pep_SUITE.logs/run.2018-06-13_17.45.59/pep_suite.delayed_receive_with_sm.1838.html)
</details>

**Comment by mongoose-im** [comment]
[4918.2](https://travis-ci.org/esl/MongooseIM/jobs/392150898) / Erlang 19.3 / mysql_redis / 1ff2c0405e5e5ee8edccef4ea4c83fb171ba7976
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1925/4918/mysql_redis.19.3) / [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4918/mysql_redis.19.3/big/ct_run.test%40travis-job-0bb212be-58f9-4b7e-944f-04b470b54d69.2018-06-14_11.04.31/index.html) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4918/mysql_redis.19.3/small/ct_run.mongooseim@localhost.2018-06-14_10.57.27/lib.mongooseim.logs/run.2018-06-14_10.57.28/suite.log.html)
**OK: 2809** / Failed: 0 / **User-skipped: 222** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[4924.1](https://travis-ci.org/esl/MongooseIM/jobs/392341358) / Erlang 19.3 / internal_mnesia / 8597a0f74f09ae6c79200958c63167633aadc6f6
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1925/4924/internal_mnesia.19.3) / [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4924/internal_mnesia.19.3/big/ct_run.test%40travis-job-46d316fb-ec59-49da-84c2-0d2040d52c40.2018-06-14_17.27.22/index.html) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4924/internal_mnesia.19.3/small/ct_run.mongooseim@localhost.2018-06-14_17.20.15/lib.mongooseim.logs/run.2018-06-14_17.20.16/suite.log.html)
**OK: 1068** / **Failed: 2** / **User-skipped: 41** / Auto-skipped: 0 
<details><summary>mod_ping_SUITE:server_ping:server_ping_pang</summary>


```erlang
{error,{{badmatch,[{[<<"localhost">>,user_ping_timeout],
          {expected_diff,1},
          {before_story,0},
          {after_story,0}}]},
    [{escalus_mongooseim,post_story_check_metrics,1,
               [{file,"/home/travis/build/esl/MongooseIM/big_tests/_build/default/lib/escalus/src/escalus_mongooseim.erl"},
                {line,74}]},
     {escalus_mongooseim,maybe_check_metrics_post_story,1,
               [{file,"/home/travis/build/esl/MongooseIM/big_tests/_build/default/lib/escalus/src/escalus_mongooseim.erl"},
                {line,51}]},
     {escalus_story,story,4,
            [{file,"/home/travis/build/esl/MongooseIM/big_tests/_build/default/lib/escalus/src/escalus_story.erl"},
             {line,75}]},
     {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1529}]},
     {test_server,run_test_case_eval1,6,
            [{file,"test_server.erl"},{line,1045}]},
     {test_server,run_test_case_eval,9,
            [{file,"test_server.erl"},{line,977}]}]}}
```

[Report log](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4924/internal_mnesia.19.3/big/ct_run.test%40travis-job-46d316fb-ec59-49da-84c2-0d2040d52c40.2018-06-14_17.27.22/big_tests.tests.mod_ping_SUITE.logs/run.2018-06-14_17.37.39/mod_ping_suite.server_ping_pang.879.html)
</details>
<details><summary>mod_ping_SUITE:server_ping_kill:server_ping_pang</summary>


```erlang
{error,{{badmatch,true},
    [{mod_ping_SUITE,check_connection,2,
             [{file,"mod_ping_SUITE.erl"},{line,153}]},
     {mod_ping_SUITE,'-server_ping_pang/1-fun-0-',2,
             [{file,"mod_ping_SUITE.erl"},{line,144}]},
     {escalus_story,story,4,
            [{file,"/home/travis/build/esl/MongooseIM/big_tests/_build/default/lib/escalus/src/escalus_story.erl"},
             {line,72}]},
     {test_server,ts_tc,3,[{file,"test_server.erl"},{line,1529}]},
     {test_server,run_test_case_eval1,6,
            [{file,"test_server.erl"},{line,1045}]},
     {test_server,run_test_case_eval,9,
            [{file,"test_server.erl"},{line,977}]}]}}
```

[Report log](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4924/internal_mnesia.19.3/big/ct_run.test%40travis-job-46d316fb-ec59-49da-84c2-0d2040d52c40.2018-06-14_17.27.22/big_tests.tests.mod_ping_SUITE.logs/run.2018-06-14_17.37.39/mod_ping_suite.server_ping_pang.891.html)
</details>

**Comment by mongoose-im** [comment]
[4926.5](https://travis-ci.org/esl/MongooseIM/jobs/392638278) / Erlang 19.3 / elasticsearch_and_cassandra_mnesia / 5129525227cb304bc036815de9095de8a1201c9b
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1925/4926/elasticsearch_and_cassandra_mnesia.19.3) / [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4926/elasticsearch_and_cassandra_mnesia.19.3/big/ct_run.test%40travis-job-c365cae2-6767-4cf4-940f-a1d55e8a7e17.2018-06-15_10.04.50/index.html) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4926/elasticsearch_and_cassandra_mnesia.19.3/small/ct_run.mongooseim@localhost.2018-06-15_09.57.46/lib.mongooseim.logs/run.2018-06-15_09.57.46/suite.log.html)
**OK: 445** / Failed: 0 / **User-skipped: 8** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[4930.2](https://travis-ci.org/esl/MongooseIM/jobs/392692140) / Erlang 19.3 / mysql_redis / 4d22a0e7dd2c7808067a693cd9744717e6938fd4
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1925/4930/mysql_redis.19.3) / [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4930/mysql_redis.19.3/big/ct_run.test%40travis-job-79b049fb-424a-4d32-85ca-2f3dabe0c321.2018-06-15_13.13.56/index.html) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4930/mysql_redis.19.3/small/ct_run.mongooseim@localhost.2018-06-15_13.06.36/lib.mongooseim.logs/run.2018-06-15_13.06.36/suite.log.html)
**OK: 2809** / Failed: 0 / **User-skipped: 222** / Auto-skipped: 0

File: big_tests/tests/pep_SUITE.erl [hunk 1]
```diff
@@ -322,3 +365,43 @@ caps_hash(PEPNodeNS) ->
...
+    ToJid = escalus_client:short_jid(To),
+    Stanza = escalus_stanza:presence_direct(ToJid, Type),
+    escalus_client:send(From, Stanza).
+
+make_friends(Bob, Alice) ->
+    % makes uni-directional presence subscriptions
+    % returns stanzas received finally by the inviter
+    send_presence(Bob, <<"subscribe">>, Alice),
+    send_presence(Alice, <<"subscribed">>, Bob),
+    escalus:wait_for_stanzas(Alice, 10, 200),
```
[the remark this answers is not in the document]
Comment (bartekgorny) [reply big_tests/tests/pep_SUITE.erl]: Perhaps it is hackish, I didn't have a better idea - the trouble is that the order of those messages is undetermined, and I have to find the delayed message, discarding everything else.

[the remark this answers is not in the document]
Comment (fenek) [reply big_tests/tests/pep_SUITE.erl]: Fair enough. Please fix the broken indent inside filter fun and resolve the conflict with master. :)

**Comment by mongoose-im** [comment]
[4973.2](https://travis-ci.org/esl/MongooseIM/jobs/397248275) / Erlang 19.3 / internal_mnesia / 567ffd0093c6da7c9dd01abc40c94dd7b8c985b4
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1925/4973/internal_mnesia.19.3) / [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4973/internal_mnesia.19.3/big/ct_run.test%40travis-job-00282621-6b9a-4b45-acae-8c95955e7810.2018-06-27_09.44.32/index.html) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/4973/internal_mnesia.19.3/small/)
**OK: 1062** / Failed: 0 / **User-skipped: 41** / Auto-skipped: 0

**Comment by michalwski** [comment]
@bartekgorny could you rebase the code or resolve the conflicts in other way? (rebase more preferable)

**Comment by mongoose-im** [comment]
[5183.1](https://travis-ci.org/esl/MongooseIM/jobs/405248206) / Erlang 19.3 / small_tests / 3b91acaeb02d25283d5ea348467ffbae32c7b888
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/1925/5183/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/1925/5183/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-07-18_09.04.06/lib.mongooseim.logs/run.2018-07-18_09.04.07/suite.log.html)

**Comment by michalwski** [comment]
@fenek are you ok with the changes now? The PR is ready to be merged, I solved conflicts and travis is green for all the jobs.

**Pull request merged**